### ISSUE TITLE
CustomPainter.hitTest missing size argument

### ISSUE THREAD

**Issue by spkersten** [flutter/flutter#28206]
`CustomPainter.hitTest` is missing the size of the widget, making it impossible to specify custom hit test behaviour when `paint` uses its `size` argument.

Simplistic example:

```dart
class CirclePainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    final center = size.center(Offset.zero);
    final borderPaint = Paint()
      ..color = const Color(0xffffffff)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 2;
    canvas.drawCircle(center, size.width/2, borderPaint);
  }

  @override
  bool hitTest(Offset position) {
    final size = ???;
    final center = size.center(Offset.zero);
    return (position - center).distance < 0.5;
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => false;
}
```

**Comment by zoechi** [comment]
Please add the output of `flutter doctor -v`.

**Comment by goderbauer** [comment]
`CustomPainter.hitTest` should indeed take a `Size` argument. That was an oversight to not provide one. I added this to the list in https://github.com/flutter/flutter/issues/24722.

Since this is a breaking change we need to carefully think about how we fix this as we do not want to annoy our users with constant breaking changes that they have to apply to their code base manually. I believe we have three options here:

1. **Make the breaking change and add the additional argument.** This will require all people subclassing CustomPainter to update their subclass. Luckily, this change can be automated and in order to assist our users with breaking changes we would like to provide them with some kind of a tool that automatically upgrades their code base. We need to figure out what infrastructure we need for that and build that for this option to work. This would leave us with a nice and concise API.

2. **Convince the Dart language team that methods in subclasses do not have to take all the arguments specified in the superclass.** This option would make your proposed change a non-breaking change because we can add `Size` to the CustomPainter without requiring any changes to subclasses. This option will also make future changes of this type easier (I'd anticipate we will have more like this...). I've filed https://github.com/dart-lang/sdk/issues/36027 to discuss this with the language team. Feel free to add your thoughts there.

3. **We make this change non-breaking by adding a new method.** This method would take an `Offset` and `Size`. By default, it would just call the existing `hitTest` with just the `Offset` ensuring that existing implementations continue to work. Subclasses, who want access to `Size` can override this new method instead. This option would require that we can come up with a sensible name for this new method, that blends in with the general API of `CustomPainter` and doesn't look like it was just an afterthought. So far, I haven't found a great name...

/cc @dnfield @Hixie

**Comment by dnfield** [comment]
My own preference would be for option 3.  I think we could then mark `hitTest` as `@deprecated` to encourage people to override the new method.

**Comment by Hixie** [comment]
The problem is what would you call it. `hitTest` is really the only reasonable name for this method.

**Comment by dnfield** [comment]
`sizedHitTest`?

If we can get the language change that would be great.  My concern with a breaking change on this is that it seems likely to break packages that will be challenging to fix for consumers - not to mention creating a need to fork the package for any further fixes.

Another option is to make CustomPaint generic and test that it takes the current interface or the new one...

**Comment by matanlurey** [comment]
@Hixie @dnfield Have you considered either a separate interface or an arity test?

```dart
abstract class CustomPainter {
  void hitTest(Offset offset);
}

class Offset {}

abstract class CustomPainterWithSize implements CustomPainter {
  @override
  void hitTest(Offset offset, {Size size});
}

class Size {}

void example(CustomPainter p) {
  // Idea 1. Check for "CustomPainterWithSize".
  if (p is CustomPainterWithSize) {
    // ...
  }
  
  // Idea 2. Check for arity.
  final hitTest = p.hitTest;
  if (hitTest is void Function(Offset, {Size size})) {
    // ... 
  }
}
```

/cc @jonahwilliams

**Comment by jonahwilliams** [comment]
There are only two places where we actually need to call the CustomPainter.hitTest - both in RenderCustomPaint

**Comment by goderbauer** [comment]
It feels like adding another interface would add unnecessary complexity. Now as a user you'd have to decide do I just want to subclass a regular CustomPainter or do I care about Size and what does size mean here? You'd have to make that decision even if you don't need to implement hitTest, so this would affect even more users.

I'd probably prefer adding a new method to the existing interface over that because that keeps the decision "do I care about size" local to when you need to override hitTest and doesn't require you to think about it when you subclass the custom painter in the first place...

**Comment by spkersten** [comment]
Wasn't the idea of the new interface to also deprecate the old one? In that case there is no choice for the user.

**Comment by goderbauer** [comment]
Doing the deprecation on the interface level would affect every user who is subclassing CustomPainter. Doing it on the method level would only effect people overriding hitTest, which I suspect would be fewer.

**Comment by dnfield** [comment]
I suspect something like this will be our cleanest, least breaking option (that doesn't require a language change):

```
abstract class CustomPainter {
  ...

  @Deprecated('use hitTestForBox')
  bool hitTest(Offset offset) => null;
  bool hitTestForBox(Offset offset, Size boxSize) => hitTest(offset); // ignore: deprecated_member
}
```

**Comment by Hixie** [comment]
I don't think we would design the API in any of those ways from the beginning. Sure, if we're willing to have a suboptimal API then there's lots of easy solutions to keep this backwards compatible.

**Comment by matanlurey** [comment]
@goderbauer:

> It feels like adding another interface would add unnecessary complexity.

Right. But the alternatives are:

* A breaking change (seems undesirable at this point)
* An arity test with `is void Function(...)` which won't produce static hints for the user
* Waiting for the Dart language to drastically change semantics or add new features

@goderbauer:

> I'd probably prefer adding a new method to the existing interface

@dnfield:

> I suspect something like this will be our cleanest, least breaking option

That is a breaking change unless you force all customers to `extend CustomPainter` (instead of `implements CustomPainter`) - AFAIK there is no way currently to do this (though there really should be, maybe a `@doNotImplement` in `package:meta`).

@Hixie:

> I don't think we would design the API in any of those ways from the beginning.

Isn't this entire conversation around trade-offs because the right API wasn't chosen? :sob:

**Comment by matanlurey** [comment]
There is another alternative, you could add a static variable that has the "last" size :crying_cat_face: 

```dart
class CustomPainter {
  /// Set before invoking [hitTest].
  static Size get hitTestSize => _hitTestSize;
}
```

**Comment by Hixie** [comment]
The question is how do we get to an ideal API. We sometimes can do it without breakage, sometimes not.

In this particular case, I think the automatic mechanical addition of the argument to legacy code, or the adding of a feature to the language to make arguments optional on implementations, are the best way to go. All the options for adding backwards compatible APIs seem like they would make the API worse.

**Comment by goderbauer** [comment]
The static variable doesn't work in the case where the same CustomPainter instance is reused in multiple places, which is legal. If it's not used in multiple places, you can just hold on to the size given in the call to paint...

**Comment by dnfield** [comment]
So one other option is to just implement this using a [LeafRenderObjectWidget](https://docs.flutter.io/flutter/widgets/LeafRenderObjectWidget-class.html).

@Hixie - Would a mechanical solution work with packages that use this?  

@matanlurey I'm not seeing why it's a breaking change...  Can you elaborate?

**Comment by matanlurey** [comment]
@Hixie:

> ... Automatic mechanical addition of the argument to legacy code

Correct me if I am wrong, but this is not iron proof. You can easily use a plugin (or package) that has not run the migration script, and if you don't control upload rights for that package, you now have to fork every transitive dependency until they run the migration script.

To be clear, this is a good idea. It is still a tough breaking change though, and unlike the other ideas (new interfaces, arity checks, static variables) there is no migration period where both APIs work, which I think is what you need.

> ... adding of a feature to the language to make arguments optional on implementations.

This seems extremely unlikely given all of the directions of the language today.

@goderbauer:

> The static variable doesn't work in the case where the same CustomPainter instance is reused

I guess I did not explain, I just mean _before_ calling `CustomPainter.hitTest` you'd set that variable. It would be a wonky API, sure, but it would let you get a fix/workaround out fast in a non-breaking way while you figure out what final API you'd like.

@dnfield:

> I don't think we've been avoiding adding methods to classes.

If that is the case, then the best thing to do, I agree, is add a second method.

**Comment by Hixie** [comment]
I don't think we should ship a "wonky" API. We have survey results that show that 80% of our user base prefer us to have a clean API over avoiding breaking changes. IMHO this means we should make sure that all our APIs are what we consider to be good clean APIs. We should certainly avoid breaking changes if we can do so without compromising API quality, but here I don't see how we can.

This particular API is intended as an interface as well as a superclass, so adding a method is definitely breaking, not just theoretically so.

The migration path is that people would update their packages' minimum required Flutter version, and people would remain on older Flutter versions until all their packages have migrated.

**Comment by willlarche** [comment]
Do we have enough consensus for a plan forward?

**Comment by JulianAssmann** [comment]
Are there any updates on this? I really need the size to calculate my hitbox. Are there currently any workarounds?

**Comment by spkersten** [comment]
A workaround would be to copy the `CustomPaint` widget and apply the change in my rejected PR that is refer to above. (If you don't need all functionality, you could also make a simpler `RenderObject` that does your painting.)

**Comment by dnfield** [comment]
IMO CustomPaint is a trap.  It tries to make something that's already not terribly hard easier, and in doing so it hides some complexity like this that you fall into later. Using a `LeafRenderObjectWidget` or a `SingleChildRenderObjectWidget` is not much more complicated (or much more code).

I think we should close this issue.  There is unfortunately no good way to fix this API without massively breaking it.  If anything we should just document it as such.

**Comment by Hixie** [comment]
Under our new breaking change policy we should be able to just add the `size` argument to `hitTest`.

**Comment by dnfield** [comment]
I'm in no way opposed to breaking this, but I'm not particularly interested in doing it myself - mainly because I strongly suspect that this really just one of many issues that people will run into as they get into more "advanced" use cases of CustomPainter, where they really likely want to just switch to using a `RenderObjectWidget` of some sort. For example, here's how the original question might be rewritten:

```dart
class CirclePainter extends LeafRenderObjectWidget {
  const CirclePainter({Key key}) : super(key: key);

  @override
  RenderObject createRenderObject(BuildContext context) => RenderCirclePainter();
}

class RenderCirclePainter extends RenderBox {
  @override
  bool get sizedByParent => true;

  @override
  void performResize() => size = constraints.biggest;

  @override
  void paint(PaintingContext context, Offset offset) {
    final center = size.center(Offset.zero);
    final borderPaint = Paint()
      ..color = const Color(0xffffffff)
      ..style = PaintingStyle.stroke
      ..strokeWidth = 2;
    context.canvas.drawCircle(center, size.width/2, borderPaint);
  }

  @override
  bool hitTestSelf(Offset position) {
    final center = size.center(Offset.zero);
    return (position - center).distance < 0.5;
  }
}
```

I think for now we should open a PR that documents this for `hitTest` on `CustomPaint` and `CustomPainter`.

**Comment by Hixie** [comment]
You'd have to also implement the intrinsic sizing functions and the baseline alignment function, iirc.

**Comment by dnfield** [comment]
Not if you just defer sizing to the parent.

**Comment by Hixie** [comment]
RenderBox just returns zero by default (and null for the baseline logic, though that's probably ok if you don't have text).

**Comment by dnfield** [comment]
Right - but if you just wrap it with a parent that takes care of these things for you, e.g. `SizedBox(height: 100, width: 100, child: CirclePainter())` all the intrinsics are taken care of by the SizedBox.

**Comment by Hixie** [comment]
I don't think that really counts as implementing the RenderBox protocol. :-)

**Comment by JulianAssmann** [comment]
My current workaround is to just save the size in the `paint` method and access it in the `hitTest ` method

```
class YourCustomPainter extends CustomPainter {
  Size _currentSize;

  @override
  void paint(Canvas canvas, Size size) {
    _currentSize = size;
  }

  @override
  bool hitTest(Offset position) {
    // Do whatever with the _currentSize here
    print('_currentSize: ' + _currentSize.width.toString() + ', ' + _currentSize.height.toString());
    return super.hitTest(position);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}
```

I don't know, if this works in all circumstances, but in my specific situation it's working fine. Maybe it helps someone.

**Comment by spkersten** [comment]
@JulianAssmann The `CustomPainter` is a configuration (like widgets) and might be used by multiple render objects, each with their own size. So the hit testing would then uses the size of the last one what was painted, instead of the one that was hit.

**Comment by fischerscode** [comment]
> Make the breaking change and add the additional argument

Alternatively a new type extending `Offset` (I called it `OffsetInBox`. The name is obviously open for debate.) could be added.
From this two new options rise:
1. `CustomPainter`s `hitTest` method takes a `OffsetInBox`. (I prefer this option)
    - Developers overwriting `hitTest` don't have to change their implementation. ✅ 
    - Developers calling `hitTest` on an `CustomBuilder` have to supply a `OffsetInBox` 💥 (But I'm not sure if anyone does that.)
    - Clean API ✅ 
2. `CustomPainter`s `hitTest` still takes a `Offset`, but `RenderCustomPaint` always calls it with an `OffsetInBox`.
    - Developers overwriting `hitTest` don't have to change their implementation. ✅ 
    - Developers calling `hitTest` on an `CustomBuilder` don't have to change their implementation. ✅ 
    - Developers overwriting `hitTest` and need the size have to check if the offset is an `OffsetInBox`. 😐 
    - Less clean API 😐 

A **minimal** example for an `OffsetInBox` implementation.
```dart
class OffsetInBox extends Offset {
  const OffsetInBox({
    required double dx,
    required double dy,
    required this.height,
    required this.width,
  }) : super(dx, dy);

  final double height;
  final double width;
}
```

In case there is positive feedback, I'll provide a PR.

**Comment by Hixie** [comment]
I think the right solution here is just for us to add a `Size size` parameter to `hitTest`, combined with same automated migration. If `dart fix` can do it then that's the obvious path, now that we have that tool (it didn't exist when this bug was originally filed). I'm not sure if `dart fix` can add a new parameter to an overridden method like this? (cc @Piinks)

**Comment by Piinks** [comment]
> I'm not sure if dart fix can add a new parameter to an overridden method like this?

I think it can! If it is a change that needs to be made to the base and all subclasses, they each need to have an individual fix written.

**Comment by Hixie** [comment]
I'm not sure I understand if that means it's possible or not. :-) The situation here is we would need every subclass of CustomPainter (in user code) to have the `hitTest` method gain a new argument.

**Comment by fischerscode** [comment]
I've looked at the dart fix documentation, and while I did not try it yet, I think it's possible.

**Comment by Piinks** [comment]
> I'm not sure I understand if that means it's possible or not.

Apologies for not being clear. I don't know.

I think it should work. You can alter a method on a given class using the `method:` key, and the change you would want to make is `addParameter`. While there may not be an existing fix rule that matches this exact use case, we have many similar cases if you search for those keywords in `flutter/packages/flutter/lib/fix_data`. As the only current customer for `dart fix`, we are still finding cases that do not work or do not have full support. If that be the case, we can file an issue with the dart team. Happy to assist on a PR.

**Comment by definev** [comment]
any progress on this?

**Comment by benthillerkus** [comment]
Please just do a breaking change, that's 1K easier than waiting 5 years for a fix

**Comment by colbymaloy** [comment]
> `CustomPainter.hitTest` should indeed take a `Size` argument. That was an oversight to not provide one. I added this to the list in #24722.

@goderbauer i feel like its massive oversight for hitTest to not have gesture information. maybe im missing something but when im using an interactiveviewer with custompainter, panning will trigger hit test method, whereas i only need it to trigger on TAP. 

think a google maps like experience. im painting items on top of a map. how do i interact with them? how do i paint 200 items performantly?
